runtime/wasm/src/host.rs

```diff
@@ -267,13 +319,243 @@ impl RuntimeHost {
                 )
             })
     }
+
+    fn handler_for_call(&self, call: &Arc<EthereumCall>) -> Result<MappingTransactionHandler, Error> {
+        // First four bytes of the input for the call are the first four bytes of hash of the function signature
+        if call.input.0.len() < 4 {
+            return Err(format_err!("Ethereum call has input with less than 4 bytes"))
+        }
+        let target_method_id = &call.input.0[..4];
+        self.data_source_call_handlers
+            .iter()
+            .find(move |handler| {
+                let fhash = keccak256(handler.function.as_bytes());
+                let actual_method_id = [fhash[0], fhash[1], fhash[2], fhash[3]];
+                target_method_id == actual_method_id
+            })
+            .cloned()
+            .ok_or_else(|| {
+                format_err!(
+                    "No call handler found for call in data source \"{}\"",
+                    self.data_source_name,
+                )
+            })
+    }
+
+    fn handler_for_block(&self) -> Result<MappingBlockHandler, Error> {
+        self
+            .data_source_block_handler
+            .clone()
+            .ok_or_else(|| {
+                format_err!(
+                    "RuntimeHost does not have a block handler in data source \"{}\"",
+                    self.data_source_name,
+                )
+            })
+    }
 }
 
 impl RuntimeHostTrait for RuntimeHost {
     fn matches_log(&self, log: &Log) -> bool {
         self.matches_log_address(log) && self.matches_log_signature(log)
     }
 
+    fn matches_call(&self, call: &EthereumCall) -> bool {
+        self.matches_call_address(call) && self.matches_call_function(call)
+    }
+
+    fn matches_block(&self, block: &EthereumBlock, call: &EthereumCall) -> bool {
+        self.data_source_block_handler.is_some() && self.matches_call_address(call)
+    }
+
+    fn process_call(
```

Comment:
Should be able to make `process_call`, `process_log`, and `process_block` more DRY